File: Leetcode/207. Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> dependency(numCourses,0);
        vector<vector<int>> adj(numCourses);
        for(auto& p:prerequisites){
            adj[p[1]].push_back(p[0]);
            dependency[p[0]]++;
        }
        queue<int> q;
        for(int i=0;i<numCourses;i++){
            if(dependency[i]==0)    q.push(i);
        }
        while(!q.empty()){
            int t=q.front();
            q.pop();
            for(int i:adj[t]){
                dependency[i]--;
                if(dependency[i]==0)   q.push(i);
            }
        }
        for(int i:dependency){
            if(i!=0)    return false;
        }
        return true;
    }
};

File: Leetcode/207. Course Schedule.py
from collections import deque
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        d=[0]*numCourses
        adj = [[] for _ in range(numCourses)]
        for i in prerequisites:
            adj[i[1]].append(i[0])
            d[i[0]]+=1
        q=deque()
        for i in range(numCourses):
            if d[i]==0:
                q.append(i)
        while q:
            t=q[0]
            q.popleft()
            for i in adj[t]:
                d[i]-=1
                if(d[i]==0):
                    q.append(i)
        for i in d:
            if i!=0:
                return False
        return True
